File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.java
class Solution {
    public int furthestBuilding(int[] heights, int bricks, int ladders) {
        int n = heights.length;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for (int i = 0; i < n - 1; i++) {
            int curr_diff = heights[i + 1] - heights[i];
            if (curr_diff > 0) {
                pq.add(curr_diff);
            }
            
            if (pq.size() > ladders) {
                bricks -= pq.poll();
            }
            
            if (bricks < 0) {
                return i;
            }
        }
        
        return n - 1;
    }
}

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        int n = heights.size();
        priority_queue<int> pq;
        
        for(int i = 0; i < n - 1; i++) {
            int curr_diff = heights[i + 1] - heights[i];
            if (curr_diff > 0) {
                pq.push(-curr_diff);
            }
            
            if (pq.size() > ladders) {
                bricks += pq.top();
                pq.pop();
            }
            
            if (bricks < 0) return i;
        }
        
        return n - 1;
    }
};